Source/Evolve/WarpXEvolve.cpp

```diff
@@ -235,9 +242,12 @@ WarpX::Evolve (int numsteps)
         }
 
         // TODO: move out
+        bool const end_of_step_loop = (step == numsteps_max - 1) || (cur_time + dt[0] >= stop_time - 1.e-3*dt[0]);
         if (evolve_scheme == EvolveScheme::Explicit) {
-            // At the end of last step, push p by 0.5*dt to synchronize
-            if (cur_time + dt[0] >= stop_time - 1.e-3*dt[0] || step == numsteps_max-1) {
+            // At the end of step loop, push p by 0.5*dt to synchronize
+            // This synchronization is not at the correct place since it is done before the window is moved,
+            // before particles are scraped, and before the electrostatic field update
```

Comment:
Should this be left here for a future PR to address, or should this be handled before merging this PR?

Reply:
I will let @dpgrote chime in, but I want to report one of his comments from February 2023:

> The buggy behavior regarding particle bounds is not dependent on input options, but will always happen. For the other, Python with electrostatic, this should just be fixed by removing this block of code and using the one below instead. I can do this in a separate PR because it will break a number of CI tests.

If the "Python with electrostatic" fix is a small bug fix (not too clear to me at this time), maybe it would be efficient to do it right away in this PR and update whatever checksum bechmarks need to be updated.

Reply:
My thought was that this PR is adding a new feature and should stand on its own. I wanted to make sure that this PR didn't break anything. The other changes would be in a separate PR since it changes results and breaks CI tests. It's a cleaner separation this way.

Reply:
@dpgrote 
Understood, that works for me.

@roelof-groenewald 
If this PR looks good to you, please feel free to go ahead and merge. We can also touch base with everyone briefly in the PR triage meeting before going ahead with the merge.

Reply:
I agree that the bug fix would be better for a follow-up PR. In that PR it would be good to also make synchronizing the default.

Another thing that occurred to us is that the particles that are scraped will not have synchronized velocities and positions. This can also be addressed in a follow-up PR.

I will make issues for both of these points.